File: maki-code-index/src/index/markdown.rs
use tree_sitter::Node;

use super::common::{LanguageExtractor, Section, SkeletonEntry, compact_ws, node_text};

pub(crate) struct MarkdownExtractor;

impl MarkdownExtractor {
    fn heading_level(node: Node, source: &[u8]) -> usize {
        match node.kind() {
            "atx_heading" => {
                // Count the # characters at the start of the raw text
                let text = node_text(node, source);
                text.chars().take_while(|&c| c == '#').count()
            }
            "setext_heading" => {
                let text = node_text(node, source);
                // `===` is level 1, `---` level 2 and other levels don't exist
                if text.trim_end().ends_with('=') { 1 } else { 2 }
            }
            _ => unreachable!(),
        }
    }

    fn collect_all_headings<'a>(
        entries: &mut Vec<(usize, SkeletonEntry)>,
        node: Node<'a>,
        source: &'a [u8],
    ) {
        if (node.kind() == "atx_heading" || node.kind() == "setext_heading")
            && let Some(heading_content) = node.child_by_field_name("heading_content")
        {
            let level = Self::heading_level(node, source);
            let text = node_text(heading_content, source).trim();
            // Prefix with level `#` to format like an atx heading
            let text = format!("{:#<1$} {text}", "", level);
            let compacted = compact_ws(&text).into_owned();
            let entry = SkeletonEntry::new(Section::Heading, node, compacted);
            entries.push((level, entry));
        }

        // Recurse into child nodes
        let mut cursor = node.walk();
        for child in node.children(&mut cursor) {
            Self::collect_all_headings(entries, child, source);
        }
    }
}

impl LanguageExtractor for MarkdownExtractor {
    fn extract_nodes_from_root(
        &self,
        root: Node,
        source: &[u8],
    ) -> (Vec<SkeletonEntry>, Vec<usize>) {
        let mut entries: Vec<(usize, SkeletonEntry)> = Vec::new();

        // Walk the entire tree to find all headings in document order
        Self::collect_all_headings(&mut entries, root, source);

        // For all entries, calculate the end position by finding the next entry
        // with the same or lower level, e.g. a `##` extends until the next `##`
        // or `#`, whichever happens first.
        //
        // If there is none then it extends until the end of the document.
        let doc_end = root.end_position().row + 1;
        for i in 0..entries.len() {
            entries[i].1.line_end = entries[i + 1..]
                .iter()
                .find(|(lvl, _)| *lvl <= entries[i].0)
                .map(|(_, e)| e.line_start - 1)
                .unwrap_or(doc_end);
        }

        let entries = entries.into_iter().map(|(_, e)| e).collect();

        (entries, Vec::new())
    }

    fn import_separator(&self) -> &'static str {
        ""
    }
}

File: plugins/index/lang/markdown.lua
return function(U)
  local get_text = U.get_text
  local compact_ws = U.compact_ws
  local new_entry = U.new_entry
  local format_skeleton = U.format_skeleton
  local SECTION = U.SECTION

  local function heading_level(node, source)
    if node:type() == "atx_heading" then
      local text = get_text(node, source)
      local count = 0
      for i = 1, #text do
        if text:sub(i, i) == "#" then count = count + 1 else break end
      end
      return count
    else
      local text = get_text(node, source)
      return text:match("=+%s*$") and 1 or 2
    end
  end

  local function collect_all_headings(entries, node, source)
    local kind = node:type()
    if kind == "atx_heading" or kind == "setext_heading" then
      local content_node = node:field("heading_content")[1]
      if content_node then
        local level = heading_level(node, source)
        local text = get_text(content_node, source):match("^%s*(.-)%s*$")
        local prefix = string.rep("#", level)
        local entry = new_entry(SECTION.Heading, node, compact_ws(prefix .. " " .. text))
        entries[#entries + 1] = { level = level, entry = entry }
      end
    end
    for _, child in ipairs(node:children()) do
      collect_all_headings(entries, child, source)
    end
  end

  return {
    extract = function(source, root)
      local tagged = {}
      collect_all_headings(tagged, root, source)

      local doc_end = root:end_() + 1
      for i, item in ipairs(tagged) do
        local end_line = doc_end
        for j = i + 1, #tagged do
          if tagged[j].level <= item.level then
            end_line = tagged[j].entry.line_start - 1
            break
          end
        end
        item.entry.line_end = end_line
      end

      local entries = {}
      for _, item in ipairs(tagged) do
        entries[#entries + 1] = item.entry
      end

      return format_skeleton(entries, {}, nil, "")
    end,
  }
end
